Spigot-Server-Patches/0424-Rate-limit-packets-incoming-from-players.patch

```diff
@@ -0,0 +1,104 @@
+From 5ee916e3f0513eb0f60718f931414278568cf55a Mon Sep 17 00:00:00 2001
+From: Spottedleaf <Spottedleaf@users.noreply.github.com>
+Date: Mon, 11 Mar 2019 12:18:29 -0700
+Subject: [PATCH] Rate limit packets incoming from players
+
+Packets are ratelimited through an interval of time to make sure latency
+spikes don't kill clients. The default limit is 250 packets/second, which
+is also far greater than the max ~70 packets/second vanilla will do according
+to my testing.
+
+Without these ratelimits in place I can happily send 40,000 move packets per
+second which will take down the server.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+index a7673dd49..7c4b45522 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -462,4 +462,11 @@ public class PaperConfig {
+         maxBookPageSize = getInt("settings.book-size.page-max", maxBookPageSize);
+         maxBookTotalSizeMultiplier = getDouble("settings.book-size.total-multiplier", maxBookTotalSizeMultiplier);
+     }
++
++    public static int packetRateLimit = 250; // per second
++    public static String packetRateLimitKickMessage = "Sent too many packets";
++    private static void packetRateLimit() {
++        packetRateLimit = Math.max(-1, getInt("settings.spam-limiter.packets-per-second", packetRateLimit));
++        packetRateLimitKickMessage = getString("messages.kick.packet-spam", packetRateLimitKickMessage);
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
+index 30f646e42..de3fb824a 100644
+--- a/src/main/java/net/minecraft/server/NetworkManager.java
++++ b/src/main/java/net/minecraft/server/NetworkManager.java
+@@ -153,6 +153,11 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
+     }
+ 
+     private static <T extends PacketListener> void a(Packet<T> packet, PacketListener packetlistener) {
++        // Paper start - Ratelimit packets
++        if (packetlistener instanceof PlayerConnection) {
++            ((PlayerConnection)packetlistener).rateLimitPacket(packet);
++        }
++        // Paper end
+         packet.a((T) packetlistener); // CraftBukkit - decompile error
+     }
+ 
+diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
+index dc8c20efb..5ef493188 100644
+--- a/src/main/java/net/minecraft/server/PlayerConnection.java
++++ b/src/main/java/net/minecraft/server/PlayerConnection.java
+@@ -141,6 +141,51 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
+         return (this.player == null) ? null : (CraftPlayer) this.player.getBukkitEntity();
+     }
+     // CraftBukkit end
++    // Paper start - ratelimit packets
++    private static final int PACKET_LIMIT_TIME_INTERVAL = 5; // seconds
```

Comment:
we expect at least 20 per second from player updates... this feels tight to be a 5s window with a default of 250.

i think 1 second windows intervals is more intuitive and friendly with a default of 150 maybe

Reply:
The 5 seconds is just the sampling interval. 

The actual limit of packets is config * interval time, see
https://github.com/PaperMC/Paper/pull/1898/files#diff-94d6881867da7acaee9d07759b8ca37cR65

Reply:
The interval is large to accommodate large latency spikes for clients, although I'm going to change the default to 150 to better target peak packet rates.